source/common/config/grpc_mux_impl.cc

```diff
@@ -107,6 +107,53 @@ void GrpcMuxImpl::sendDiscoveryRequest(absl::string_view type_url) {
   }
 }
 
+void GrpcMuxImpl::loadConfigFromDelegate(const std::string& type_url,
+                                         const std::vector<std::string>& resource_names,
+                                         ControlPlaneStats& control_plane_stats) {
+  if (!xds_resources_delegate_.has_value()) {
+    return;
+  }
+  ApiState& api_state = apiStateFor(type_url);
+  if (api_state.watches_.empty()) {
+    // No watches, so exit without loading config from storage.
+    return;
+  }
+
+  TRY_ASSERT_MAIN_THREAD {
+    std::vector<envoy::service::discovery::v3::Resource> resources =
+        xds_resources_delegate_->getResources(XdsConfigSourceId{target_xds_authority_, type_url},
+                                              resource_names);
+    if (resources.empty()) {
+      // There are no persisted resources, so nothing to process.
+      return;
+    }
+
+    std::vector<DecodedResourcePtr> decoded_resources;
+    OpaqueResourceDecoder& resource_decoder = *api_state.watches_.front()->resource_decoder_;
+    std::string version_info;
+    for (const auto& resource : resources) {
+      if (version_info.empty()) {
+        version_info = resource.version();
+      } else {
+        ASSERT(resource.version() == version_info);
+      }
+
+      decoded_resources.emplace_back(
+          std::make_unique<DecodedResourceImpl>(resource_decoder, resource));
+    }
+
+    processDiscoveryResources(decoded_resources, api_state, type_url, version_info,
+                              /*call_delegate=*/false);
+  }
+  END_TRY
+  catch (const EnvoyException& e) {
+    // TODO(abeyad): do something more than just logging the error and incrementing a counter?
+    control_plane_stats.xds_local_load_failed_.inc();
```

Comment:
This can probably be a call to delegate signaling that the config was rejected.
This way the delegate can free resources, if needed.

Reply:
Done